daemon/cluster/executor/container/adapter.go

```diff
@@ -153,6 +154,11 @@ func (c *containerAdapter) createNetworks(ctx context.Context) error {
 			if _, ok := err.(libnetwork.NetworkNameError); ok {
 				continue
 			}
+			// We will continue if CreateManagedNetwork returns forbidden error.
+			// Other callers still can treat it as Error.
+			if errdefs.IsForbidden(err) {
```

Comment:
I’ll have to give this more thinking (haven’t looked at what other situations return), but isn’t the case where we try to create a pre-defined network a _conflict_ (i.e. the network already “exists”).

Forbidden sounds like “you don’t have access to this endpoint” (but you do). Other options would be “invalid argument” (passed value is not valid)

Just dumping my thinking here, so that I don’t forget :smile:

Reply:
Conflict would imply that if the state changed then it would be possible to perform the action. In this case it is not allowed.

Reply:
Predefined networks are created when Daemon starts. When user creates service with predefined network, we know that network already exists and return saying you can't create. But service creation still can go ahead as the predefined network exists already. so we can ignore this error and continue in this scenario.

Reply:
> Conflict would imply that if the state changed then it would be possible to perform the action. In this case it is not allowed.

Right, yes. I was looking at the 403 because I wondered if it could cause issues in situations where a 403 could also occur due to authentication issues to the API. A generic `400` (bad request - the name is reserved/invalid) could've been an option for that.


Looking at [`NetworkNameError`](https://github.com/docker/libnetwork/blob/cb09a3770a365d8a4ed54f66e284041daf09ddef/error.go#L101-L108), I see that is also a "Forbidden" error; should we be more specific in this case, and check if it's actually a "forbidden" error, because it's a predefined network?

Given that we already walk different paths inside `createNetwork()` if the caller is `agent`, we could also use the `agent` option to just return a `nil`  instead;

```go
if runconfig.IsPreDefinedNetwork(create.Name) {
   if agent {
      return nil, nil
   }
    err := fmt.Errorf("%s is a pre-defined network and cannot be created", create.Name)
    return nil, errdefs.Forbidden(err)
}
```